### PR TITLE
LP-1709  Integrate badge progress with my badges and trophycase

### PR DESCRIPTION
**Pull request by muak50**
### Description

[LP-1709](https://philanthropyu.atlassian.net/browse/LP-1709)

### Notes
Integrate badge progress with my badges and trophycase

### CODE REVIEW

File: openedx/features/badging/helpers.py [hunk 1]
```diff
@@ -194,3 +205,73 @@ def send_user_badge_notification(user, my_badge_url, badge_name):
...
+
+def get_discussion_team_ids(course_id, discussion_room_id, badges):
+    """
+    Return generic dictionary that contain another dictionary with course_id as key and contain
+    discussion_room_id and team_room_id.
+    :param course_id: Course Id in String format
+    :param discussion_room_id: Discussion room Id in Int format
+    :param badges: Dictionary contain badges
+    """
+    obj = {
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/helpers.py]: Done

File: openedx/features/badging/helpers.py [hunk 2]
```diff
@@ -194,3 +205,73 @@ def send_user_badge_notification(user, my_badge_url, badge_name):
...
+    :param badges_list: Dictionary contain badges both conversationalist badges and team badges
+    """
+    discussion_posts_count = course[course_key][DISCUSSION_COUNT_KEY]
+    conversationalist_badges = add_badge_progress(
+        badges_list[CONVERSATIONALIST[CONVERSATIONALIST_ENTRY_INDEX]], discussion_posts_count)
+    badges_list[CONVERSATIONALIST[CONVERSATIONALIST_ENTRY_INDEX]] = conversationalist_badges
+
+    team_posts_count = course[course_key][TEAM_COUNT_KEY] if TEAM_COUNT_KEY in course[course.keys()[0]] else 0
+    team_badges = add_badge_progress(badges_list[TEAM_PLAYER[TEAM_PLAYER_ENTRY_INDEX]], team_posts_count)
+    badges_list[TEAM_PLAYER[TEAM_PLAYER_ENTRY_INDEX]] = team_badges
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/helpers.py]: Grammatically both are correct. Kindly elaborate why this needs to be changed.

File: openedx/features/badging/views.py [hunk 3]
```diff
@@ -27,12 +32,37 @@ def trophycase(request):
...
+
+    discussion_courses = DiscussionCommunity.objects.filter(
+        course_id__in=course_key_list).values(COURSE_ID_KEY, COMMUNITY_URL_KEY)
+
+    courses = []
+    if discussion_courses:
+        for discussion_course in discussion_courses:
+            course_id = str(discussion_course[COURSE_ID_KEY])
+            courses.append(get_discussion_team_ids(
+                course_id, int(discussion_course[COMMUNITY_URL_KEY].split("/")[0]), trophycase_dict[course_id]))
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/views.py]: Done

File: openedx/features/badging/views.py [hunk 4]
```diff
@@ -27,12 +32,37 @@ def trophycase(request):
...
+    else:
+        ## WE are passing 0 as discussion id in case DiscussionCommunity didn't provide us with any data so that our
+        ## API won't break.
+        for course_id in trophycase_dict.keys():
+            courses.append(get_discussion_team_ids(course_id, 0, trophycase_dict[course_id]))
+
+    status_code, response = NodeBBClient().badges.get_progress(
+        request_data=get_badge_progress_request_data(user.username, courses))
+
+    if response:
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/views.py]: Done

File: openedx/features/badging/views.py [hunk 5]
```diff
@@ -27,12 +32,37 @@ def trophycase(request):
...
+        for course in response[COURSES_KEY]:
+            add_posts_data_in_badges_list(course, course.keys()[0], trophycase_dict[course.keys()[0]][BADGES_KEY])
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/views.py]: Done

[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/views.py]: Done. Good Suggestion

File: openedx/features/badging/views.py [hunk 6]
```diff
@@ -27,12 +32,37 @@ def trophycase(request):
...
+
     return render_to_response(
-            "features/badging/trophy_case.html",
-            {
-                'trophycase_data': trophycase_dict
-            }
-        )
+        "features/badging/trophy_case.html",
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/views.py]: Done

File: openedx/features/badging/views.py [hunk 7]
```diff
@@ -53,10 +83,27 @@ def my_badges(request, course_id):
...
+        discussion_id = 0
+
+    ## Here we are dealing with just one course so we can wrap this course details in '[]' because our API on NOdeBB
+    ## side only accepts params in this format.
+    courses = [get_discussion_team_ids(course_id, int(discussion_id), badges)]
+
+    status_code, response = NodeBBClient().badges.get_progress(
+        request_data=get_badge_progress_request_data(user.username, courses))
+
+    if response:
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/views.py]: Done

File: openedx/features/badging/views.py [hunk 8]
```diff
@@ -53,10 +83,27 @@ def my_badges(request, course_id):
...
+        ## Here we are dealing with just one course so we are getting Zero(0th) index from list.
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/views.py]: Done

File: openedx/features/badging/helpers.py [hunk 9]
```diff
@@ -194,3 +205,73 @@ def send_user_badge_notification(user, my_badge_url, badge_name):
...
+    :param course_badges: Badges data either conversationalist badges or team badges
+    :param posts_count: Count of post in Int format
+    """
+    previous_threshold = None
+    for index, badge in enumerate(course_badges):
+        if not previous_threshold:
+            previous_threshold = badge[THRESHOLD_LABEL_KEY]
+        if index > 0:
+            badge[THRESHOLD_LABEL_KEY] = badge[THRESHOLD_LABEL_KEY] - previous_threshold
+            previous_threshold = badge[THRESHOLD_LABEL_KEY] + previous_threshold
```
Comment (zeshanarifios) [inline openedx/features/badging/helpers.py]: I think we can simplify this as 0 doesn't effect for index 0.
```
previous_threshold = 0
for index, badge in enumerate(course_badges):
	badge[THRESHOLD_LABEL_KEY] = badge[THRESHOLD_LABEL_KEY] - previous_threshold
	previous_threshold = badge[THRESHOLD_LABEL_KEY] + previous_threshold
```

File: openedx/features/badging/views.py [hunk 10]
```diff
@@ -18,7 +23,7 @@ def trophycase(request):
 
     # Get course id and course name of courses user is enrolled in
     enrolled_courses_data = CourseEnrollment.enrollments_for_user(user).order_by(
-        'course__display_name').values_list('course_id', 'course__display_name')
+        'course__display_name').values_list(COURSE_ID_KEY, 'course__display_name')
```
Comment (zeshanarifios) [inline openedx/features/badging/views.py]: can we make it consistent.. As for one param we are using constant and for one we are using string.

File: openedx/features/badging/views.py [hunk 11]
```diff
@@ -27,12 +32,37 @@ def trophycase(request):
 
     trophycase_dict = populate_trophycase(user, enrolled_courses_data, earned_user_badges)
 
+    ## Get list of dictionary keys in CourseKey Format
+    course_key_list = [CourseKey.from_string(unicode(course)) for course in trophycase_dict.keys()]
```
Comment (zeshanarifios) [inline openedx/features/badging/views.py]: Can you please update course to `course_key` or `course_id` to make it more appropriate

File: openedx/features/badging/views.py [hunk 12]
```diff
@@ -53,10 +83,27 @@ def my_badges(request, course_id):
 
     badges = get_course_badges(user, course_key, earned_user_badges)
 
+    ## This will always get result because every course must have discussion board and this will return its ID.
+    try:
+        discussion_id = DiscussionCommunity.objects.get(course_id=course_key).community_url.split("/")[0]
```
Comment (zeshanarifios) [inline openedx/features/badging/views.py]: let's please use single quotes

File: common/lib/nodebb_client/badges.py [hunk 13]
```diff
@@ -16,3 +16,10 @@ def delete(self, badge_id):
         delete badge configuration
         """
         return self.client.delete('/api/v2/badge-config/{}'.format(badge_id))
+
+    def get_progress(self, request_data):
+        """
+        Get badges progress
+        """
+        payload = request_data
+        return self.client.post('/api/v2/badges/badge-progress', **payload)
```
[the remark this answers is not in the document]
Comment (muak50) [reply common/lib/nodebb_client/badges.py]: We cannot, this post request take '**kwargs' .

File: openedx/features/badging/views.py [hunk 14]
```diff
@@ -6,9 +6,14 @@
...
 from student.models import CourseEnrollment
+from nodebb.models import DiscussionCommunity
+from common.lib.nodebb_client.client import NodeBBClient
 
-from .helpers import populate_trophycase, get_course_badges
+from .helpers import populate_trophycase, get_course_badges, add_posts_data_in_badges_list,\
+    get_discussion_team_ids, get_badge_progress_request_data
 from .models import UserBadge
+from .constants import BADGES_KEY, COURSES_KEY, COURSE_ID_KEY, COMMUNITY_URL_KEY
+
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/views.py]: Done

File: openedx/features/badging/helpers.py (hunk 9 above)
Comment (muak50) [reply openedx/features/badging/helpers.py]: We need to assign previous_threshold as threshold of first badge so this is being doing here. the solution you suggested doesn't cater that requirement.

File: openedx/features/badging/views.py (hunk 11 above)
Comment (muak50) [reply openedx/features/badging/views.py]: Done

File: openedx/features/badging/views.py (hunk 12 above)
Comment (muak50) [reply openedx/features/badging/views.py]: Done

File: openedx/features/badging/helpers.py [hunk 15]
```diff
@@ -194,3 +205,73 @@ def send_user_badge_notification(user, my_badge_url, badge_name):
...
+    :param course: Dictionary that contain information related to posts count in Discussion room and Team room.
+    :param course_key: Course Id in String format
+    :param badges_list: Dictionary contain badges both conversationalist badges and team badges
+    """
+    discussion_posts_count = course[course_key][DISCUSSION_COUNT_KEY]
+    conversationalist_badges = add_badge_progress(
+        badges_list[CONVERSATIONALIST[CONVERSATIONALIST_ENTRY_INDEX]], discussion_posts_count)
+    badges_list[CONVERSATIONALIST[CONVERSATIONALIST_ENTRY_INDEX]] = conversationalist_badges
+
+    team_posts_count = course[course_key][TEAM_COUNT_KEY] if TEAM_COUNT_KEY in course[course.keys()[0]] else 0
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/helpers.py]: Done

File: openedx/features/badging/views.py (hunk 10 above)
Comment (muak50) [reply openedx/features/badging/views.py]: Done

File: common/lib/nodebb_client/badges.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (zeshanarifios) [reply common/lib/nodebb_client/badges.py]: Does this not work ?
```
	def get_progress(self, request_data):
        """
        Get badges progress
        """
        return self.client.post('/api/v2/badges/badge-progress', **request_data)
```

or 

```
	def get_progress(self, payload):
        """
        Get badges progress
        """
        return self.client.post('/api/v2/badges/badge-progress', **payload)
```
in this case parameter also needs to be updated where this method is called

File: openedx/features/badging/helpers.py (hunk 9 above)
Comment (zeshanarifios) [reply openedx/features/badging/helpers.py]: This line will do this for you 
```previous_threshold = badge[THRESHOLD_LABEL_KEY] + previous_threshold```
**For first iteration**
badge[THRESHOLD_LABEL_KEY] has threshold of first badge
and previous_threshold is 0, That's why it doesn't effect value when added

File: openedx/features/badging/views.py [hunk 16]
```diff
@@ -83,22 +82,22 @@ def my_badges(request, course_id):
...
     try:
-        discussion_id = DiscussionCommunity.objects.get(course_id=course_key).community_url.split("/")[0]
+        discussion_id = DiscussionCommunity.objects.get(course_id=course_key).community_url.split('/')[0]
     except DiscussionCommunity.DoesNotExist:
         discussion_id = 0
 
-    ## Here we are dealing with just one course so we can wrap this course details in '[]' because our API on NOdeBB
-    ## side only accepts params in this format.
+    # Here we are dealing with just one course so we can wrap this course details in '[]'
+    # because our API on NOdeBB side only accepts params in this format.
```
Comment (zahid-arbisoft) [inline openedx/features/badging/views.py]: Lets fis type `NOdeBB` to `NodeBB`

Comment (muak50) [reply openedx/features/badging/views.py]: Done

File: openedx/features/badging/helpers.py [hunk 17]
```diff
@@ -194,3 +205,73 @@ def send_user_badge_notification(user, my_badge_url, badge_name):
...
+    return {
+        USERNAME_KEY: username,
+        COURSES_KEY: json.dumps(courses)
+    }
+
+
+def add_posts_data_in_badges_list(course, course_key, badges_list):
+    """
+    Add post count data in conversationalist badges and team badges
+    :param course: Dictionary that contain information related to posts count in Discussion room and Team room.
```
[the remark this answers is not in the document]
Comment (muak50) [reply openedx/features/badging/helpers.py]: Its already renamed and parameters also changed. Its purpose is to add posts count data in badges dictionary.

File: openedx/features/badging/helpers.py (hunk 9 above)
Comment (muak50) [reply openedx/features/badging/helpers.py]: Done... Good suggestion

**Pull request merged**